lib/services/references.js

```diff
@@ -39,6 +39,31 @@ function listDeepObjects(obj, filter) {
   return list;
 }
 
+/**
+ * Search through an object and find all deep key-value pairs matching a filter.
+ * @param {object} obj
+ * @param {Function} [filter=_.identity] Optional filter
+ * @param {array} [path] Path of this value
+ * @returns {object}
+ */
+function listDeepValuesByKey(obj, filter = _.identity, path = []) {
```

Comment:
Unfortunately the order listDeepValuesByKey results in is different from listDeepObjects due to your use of recursion (depth vs breadth first).

I tested on an article - and this matters because if you change the default order then when people update their version and rely on a certain rendering order, their code will behave differently and potentially break.

Reply:
hey phil. tested on an article as well. the only thing that would be affected here are refs that are properties, which we already know order shouldn't be relied upon for those. array ordering is still maintained unless there is a ref inside of a ref (which afaik would never happen before data is populated)

Reply:
i'll get an example to show what i mean

Reply:
so testing against [this article](https://gist.github.com/olsonpm/ecc902d978ca5e3ef2e93e0bf85f4b92)

listDeepObjects returns with the order of [these refs](https://gist.github.com/olsonpm/5f8fcf37c6ad5cbc668714fec5647b34)
whereas listDeepValuesByKey returns with [this order](https://gist.github.com/olsonpm/4e145b14013cb50b3518af9e7212e07e)

In terms of keeping the public api of amphora stable, listDeepValuesByKey needs to preserve the ordering by default, regardless whether the order *should* be relied upon.

For instance, we were relying upon the rendering order and if that changed with a minor update, that would be a frustrating bug to find.

Ultimately it's the lib author's definition of "the api" and what constitutes semver changes so if you disagree then we can let them decide.